### PR TITLE
Various cleanups to the ABI handling code

### PR DESCRIPTION
**Pull request by bjorn3**
I originally did these in the process of adding exception support to Cranelift, but I've extracted it out that branch and rebased it as these changes should be less controversial and it will help with rebasing the exception support branch in the future.

### CODE REVIEW

**Review by abrown — COMMENTED** [review]
I read through this today and it all looks pretty reasonable. I feel like I need to review the safety bits, such as some of that register code, some more, though. Or maybe @jameysharpor @elliottt  if you want to jump in?

**Comment by jameysharp** [comment]
I'm excited to take a look next week if nobody else gets to this first!

**Comment by jameysharp** [comment]
I tried to review this and got kind of lost. I really appreciate that there are nicely separated commits, but I think this would be easier to review if those were fully separate PRs, if that's not too much trouble? It would be especially helpful if you can explain more about the motivation for each change in the commit/PR description.

**Comment by bjorn3** [comment]
Split out https://github.com/bytecodealliance/wasmtime/pull/8903. I will rebase this PR once it merges.

**Comment by bjorn3** [comment]
Rebased

**Review by cfallin — APPROVED** [review]
Looks reasonable, thanks! Just one nit below.

File: cranelift/codegen/src/machinst/abi.rs [hunk 1]
```diff
@@ -1994,7 +1992,7 @@ impl<M: ABIMachineSpec> CallSite<M> {
         sigs: &SigSet,
         sig_ref: ir::SigRef,
         extname: &ir::ExternalName,
-        opcode: ir::Opcode,
+        is_tail_call: bool,
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply cranelift/codegen/src/machinst/abi.rs]: Done

**Pull request merged**